### PR TITLE
Fixes #20295 - Sanitize webpack_directories json

### PR DESCRIPTION
**Pull request by xprazak2**

### CODE REVIEW

**Comment by ohadlevy** [comment]
is the warning messages are STDOUT or STDERR? if its ERR you might just ignore STDERR ?

I didnt try, but maybe:
require('child_process').execSync('command', {stdio:[0,1]});

**Comment by xprazak2** [comment]
I haven't yet found where it is coming from, but ignoring STDERR does not help.

**Comment by tstrachota** [comment]
@xprazak2 the output comes from here https://github.com/theforeman/foreman_openscap/blob/master/foreman_openscap.gemspec when the gem is loaded. I think it should be printed to stderr but it's a different issue.

For the means of sanitization to avoid such failures in future it should be enough to keep `puts` in `plugin_webpack_directories.rb` and always use only the last printed line, because we're sure the json will be printed last.

I wonder if it makes sense to print the cropped lines somewhere, because if they appear it's most likely an error that we shouldn't hide.

**Comment by xprazak2** [comment]
Ahh, that was probably the one place where I did not look, thanks. I'll raise an issue for warnings to go to stderr. I swapped `puts` for `print` because it does not add the newline character, so I can grab the last line directly. With `puts`, the last item is actually an empty string when I do the split:

```js
"some debug output\n{json: true}\n".split("\n") // ["some debug output", "{json: true}", ""]
```

but I can keep `puts` and handle it on js side.

We are not hiding the errors, they still appear when rails server starts.

**Comment by xprazak2** [comment]
I put the `puts` back, I also ignore stderr of the script as well as grab only the last line which is JSON we want if there are multiple lines.

**Comment by tstrachota** [comment]
A comment in the sanitize function explaining why we're actually doing this would be nice. Otherwise :+1: the code looks good and it works nicely!

File: config/webpack.config.js [hunk 1]
```diff
@@ -16,7 +16,18 @@ var production =
   process.env.RAILS_ENV === 'production' ||
   process.env.NODE_ENV === 'production';
 
-var plugins = JSON.parse(execSync(path.join(__dirname, '../script/plugin_webpack_directories.rb')));
+var sanitizeWebpackDirs = function (pluginDirs) {
```
Comment (ohadlevy) [inline config/webpack.config.js]: how about syntax like:
```js
var sanitizeWebpackDirs = function (pluginDirs) {
  var splitDirs = pluginDirs.toString().split('\n').reverse();

  return splitDirs.length > 2 ? splitDirs[1] : pluginDirs;
};

var webpackDirs = execSync(path.join(__dirname, '../script/plugin_webpack_directories.rb'), {
  stdio: ['pipe', 'pipe', 'ignore']
});
var plugins = JSON.parse(sanitizeWebpackDirs(webpackDirs));
```

**Comment by xprazak2** [comment]
I added a comment explaining what the function does and improved syntax.

**Review by dLobatog — APPROVED** [review]

**Review by dLobatog — CHANGES_REQUESTED** [review]
@xprazak2 Looks like you need to rebase to make sure the ruby2ruby changes are in, tests should pass then 😄

**Pull request merged**